### PR TITLE
[hotfix] CPU Detection 

### LINKED ISSUE #10809: [docker] / [core] Regression in detection of multi CPU count after docker fix
### What is the problem?

The Ray snapshot 1.1.0 will detect only 1 CPU on certain multi-core systems **when not using docker** due to this change from https://github.com/ray-project/ray/pull/10507/files#diff-dd52f82c5ae34621334f725acaeacc81R502 

```
def _detect_docker_cpus()
    cpu_quota_file_name = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us"
    cpu_share_file_name = "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
    num_cpus = 0
    if os.path.exists(cpu_quota_file_name) and os.path.exists(
            cpu_quota_file_name):
        with open(cpu_quota_file_name, "r") as f:
            num_cpus = int(f.read())
        if num_cpus != -1:
            with open(cpu_share_file_name, "r") as f:
                num_cpus /= int(f.read())
            return num_cpus
```

My current cgroup output is:
```
# cat /proc/self/cgroup
11:cpuset:/
10:hugetlb:/
9:perf_event:/
8:cpu,cpuacct:/
7:devices:/
6:memory:/
5:net_cls,net_prio:/
4:blkio:/
3:freezer:/
2:pids:/
1:name=systemd:/user.slice/user-0.slice/session-362.scope
```

Code was added to python/ray/utils.py which now requires an explicit environmental variable to be set to detect the number of CPUs. This is undesirable.

The code is in the area of ``` if "RAY_USE_MULTIPROCESSING_CPU_COUNT" in os.environ:```

Please provide a better more robust way of detecting whether you are in docker and do not require a generic setup to explicitly set the env variable.

I propose the following to detect docker:

```
def is_docker():
    path = '/proc/self/cgroup'
    return (
        os.path.exists('/.dockerenv') or
        os.path.isfile(path) and any('docker' in line for line in open(path))
    )
```

### PR DESCRIPTION
**Pull request by wuisawesome**
## Related issue number

Closes #10809 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/utils.py [hunk 1]
```diff
@@ -499,22 +499,55 @@ def get_system_memory():
...
+                        cpu_ids.append(int(num_or_range))
+                cpuset_num = len(cpu_ids)
+        except Exception as e:
+            logger.exception("Unexpected error calculating docker cpu quota",
+                             e)
+
+    if cpu_quota and cpuset_num:
+        return min(cpu_quota, cpuset_num)
+    else:
+        return cpu_quota or cpuset_num
```
Comment (yiranwang52) [inline python/ray/utils.py:550]: is it okay if this function return None on some platforms?

Comment (wuisawesome) [reply python/ray/utils.py:550]: good point...

updated `get_num_cpus` to check for None. IMO `None` makes sense as a return value though.

**Review by yiranwang52 — APPROVED** [review]

File: python/ray/utils.py (hunk 1 above)
Comment (yiranwang52) [reply python/ray/utils.py:550]: agree.

**Comment by richardliaw** [comment]
looks like sgd tests are failing for some reason?

File: python/ray/utils.py [hunk 2]
```diff
@@ -543,6 +576,14 @@ def get_num_cpus():
                     "`RAY_USE_MULTIPROCESSING_CPU_COUNT=1` as an env var "
                     "before starting Ray. Set the env var: "
                     "`RAY_DISABLE_DOCKER_CPU_WARNING=1` to mute this warning.")
+            # TODO (Alex): We should probably add support for fractional cpus.
+            if int(docker_count) != float(docker_count):
+                logger.warning(
+                    f"Ray currently does not support initializing Ray"
+                    f"with fractional cpus. Your num_cpus will be "
+                    f"truncated from {docker_count} to "
+                    f"{int(docker_count)}.")
```
Comment (ijrsvt) [inline python/ray/utils.py:585]: Do we want to round up?

File: python/ray/utils.py [hunk 3]
```diff
@@ -543,6 +576,14 @@ def get_num_cpus():
...
+            docker_count = int(docker_count)
```
Comment (ijrsvt) [inline python/ray/utils.py:586]: Also you need to set `cpu_count = int(docker_count)` because that is what the function returns.

**Review by ijrsvt — DISMISSED** [review]
Looks good, small problem with the return value.

**Review by richardliaw — APPROVED** [review]

**Comment by rhamnett** [comment]
Fix confirmed for my configuration

**Pull request merged**